File: sol/solution/0201-0300/0202/Solution.erl
-spec is_happy(N :: integer()) -> boolean().
square_sum(0) -> 0;
square_sum(N) ->
    D = N rem 10,
    D * D + square_sum(N div 10).
is_happy(N) -> is_happy(N, square_sum(N)).
is_happy(Slow, Fast) when Fast == 1 -> true;
is_happy(Slow, Fast) when Slow == Fast -> false;
is_happy(Slow, Fast) ->
    is_happy(square_sum(Slow), square_sum(square_sum(Fast))).

File: sol/solution/0201-0300/0202/Solution.ex
defmodule Solution do
  @spec is_happy(n :: integer) :: boolean
  def is_happy(n), do: detect(n, next(n))
  defp next(n) do
    n
    |> Integer.digits()
    |> Enum.map(&(&1 * &1))
    |> Enum.sum()
  end
  defp detect(slow, fast) when fast == 1, do: true
  defp detect(slow, fast) when slow == fast, do: false
  defp detect(slow, fast), do: detect(next(slow), next(next(fast)))
end